### ISSUE TITLE
Proper care and feeding of Request sockets

### ISSUE THREAD

**Issue by randycoulman** [rolftimmermans/zeromq-ng#44]
Thanks so much for the work you're doing on this next-gen 0MQ client.  I really like where you're going with it so far!  And the API documentation is already a significant improvement, so thank you!

I got here after running into an issue with the original library in the hopes that this new implementation would resolve the issues I was seeing.

I have an application that uses a `Request` socket to communicate with another application and exposes a GraphQL API.  Some of the queries need to make multiple requests over the socket and we were running into the case where one of the requests would just hang silently.  We think this was because we were trying to send multiple requests at the same time.

After switching to the new library, we were able to reproduce the problem, but this time we actually got an error message (`EAGAIN`, as expected), so at least now the problem is very clear and explicit.

We've tried using the `queue` example to ensure that we only have one request awaiting a response, but we're still seeing the `EAGAIN` errors.  They seem to be a bit less frequent, but they're still happening.  What's worse is that, once we get one of these errors, the socket remains in a bad state and every subsequent request gets an `EAGAIN` error.

I have more testing I want to do, but I thought I'd ask here what your recommended approach is for this kind of scenario with the new library?

For reference, here's the code I'm using.  The various GraphQL resolvers ultimately call `sendSequentially` with the message they want to send, and when one of the resolvers is `await`ing its response, another resolver might then call `sendSequentially`.

```js
  const socket = new Request();
  const queue = [];
  const maxQueueSize = 100;
  let sending = false;

  socket.connect("...address goes here...")

  const sendPendingMessages = async () => {
    if (sending) return;

    sending = true;

    while (queue.length) {
      const { message, reject, resolve } = queue.shift();

      try {
        await socket.send(message);
        const response = socket.receive();

        resolve(response);
      } catch (err) {
        reject(err);
      }
    }

    sending = false;
  };

  const sendSequentially = message => {
    if (queue.length > maxQueueSize)
      throw new Error("Too many pending messages");

    return new Promise((resolve, reject) => {
      queue.push({ message, reject, resolve });
      sendPendingMessages();
    });
  };
```

**Comment by rolftimmermans** [comment]
You should `await` the returned promise of `socket.receive()`.

Your inner loop should look like this:

```
try {
  await socket.send(message);
  const response = await socket.receive(); // <- note await here
  resolve(response);
}
```

What this does is that the next queued message is not sent until an answer is received for the previous message. This is required by the `Request`/`ZMQ_REQ` socket types. You can't send a message if the socket is waiting for a reply.

Of course this changes the behaviour of the code. You may have expected requests to be issued simultaneously with a `Request` socket type, but this is not possible (per the design of the `ZMQ_REQ` socket type in libzmq).

If your requests **should** be issued in parallel, then you can do one of the following:
* open a `Request` socket for each individual request; or
* design your communication protocol to use `Dealer` sockets instead of `Request`

**Comment by randycoulman** [comment]
Yup, you're right, thanks!  I can't believe I missed that one.  I appreciate your help!

**Comment by rolftimmermans** [comment]
No worries, hope you can get it working!

I also see there is 0 documentation regarding the different socket types in zeromq-ng, maybe that would have helped. I'll try to add something, but any contribution is welcome!

**Comment by randycoulman** [comment]
Yes, I got it working.  I looked that the upstream ZeroMQ docs regarding the socket types.  For now, a Request socket is what we want, but we'll thinking about switching to a Router or Dealer when it makes sense.  For now, we just needed to serialize the requests/responses going to the socket, and the queue works perfectly for that (now that I have the additional `await` in there).  Again, thanks for the quick response!
